rewrite-java/src/main/java/org/openrewrite/java/trait/variable/Parameter.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2023 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.trait.variable;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.openrewrite.Cursor;
+import org.openrewrite.Validated;
+import org.openrewrite.java.trait.Element;
+import org.openrewrite.java.trait.member.Callable;
+import org.openrewrite.java.tree.Flag;
+import org.openrewrite.java.tree.J;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a parameter in a {@link org.openrewrite.java.tree.J.MethodDeclaration} or
+ * {@link org.openrewrite.java.tree.J.NewClass}.
+ */
+public interface Parameter extends LocalScopeVariable {
+
+    J.VariableDeclarations.NamedVariable getNamedVariable();
+
+    /**
+     * The zero indexed position of the parameter in the method declaration.
+     */
+    int getPosition();
+
+    boolean isVarArgs();
+
+    static Validated of(Cursor c) {
```

Comment:
One problem I see with these `of(Cursor)` methods is that typically a visitor method starts by calling `xyz = super.visitXyz(xyz, ctx);` and assigns the result to some local variable. Now, if that ended up returning a different object, because it was transformed by some visitor method, then the cursor's object will be different. So either you would have to produce a new cursor like `new Cursor(cursor.getParent(), xyz)` or you would have to pass in the object into the `of()` method (something like `of(Cursor, J.VariableDeclarations.NamedVariable)`.

Reply:
> One problem I see with these `of(Cursor)` methods is that typically a visitor method starts by calling `xyz = super.visitXyz(xyz, ctx);` and assigns the result to some local variable.

I agree, but I've gotten feedback from others in the past that taking both arguments is redundant. As an example, see `InvocationMatcher.Advanced`. That API formerly took two arguments, the cursor and the `Expression`. But I was talked into only accepting the `Cursor`.

Reply:
I just quickly browsed through your PR again. Did I see it correctly that you changed to requiring a `(Cursor, Tree)` tuple?

Reply:
Not everywhere. One place in a protected method.